src/leap/mx/bounce.py

```diff
@@ -0,0 +1,524 @@
+    :type orig_msg: email.message.Message
+    :param reason: The reason for the bounce.
+    :type reason: str
+
+    :return: The bounce message.
+    :rtype: MIMEMultipartReport
+
+    :raise InvalidReturnPathError: Raised when the "Return-Path" header of the
+                                   original message is invalid for creating a
+                                   bounce message.
+    """
+    # abort creation if "Return-Path" header is invalid
+    orig_rpath = orig_msg.get("Return-Path")
+    if not _check_valid_return_path(orig_rpath):
+        raise InvalidReturnPathError
+
+    orig_to = orig_msg.get("To")
+
+    msg = MIMEMultipartReport()
+    msg['From'] = bounce_from
+    msg['To'] = orig_rpath
+    msg['Date'] = formatdate(localtime=True)
+    msg['Subject'] = bounce_subject
+    msg['Return-Path'] = "<>"  # prevent bounce message loop, see RFC 3834
+
+    # create and attach first required part
+    decoded_to = " ".join([x[0] for x in decode_header(orig_to)])
+    wrapped_reason = wrap(reason, 74)
+    wrapped_reason = map(lambda l: "    " + l, wrapped_reason)
+    wrapped_reason = "\n".join(wrapped_reason)
+    text = BOUNCE_TEMPLATE.format(decoded_to, wrapped_reason)
+    msg.attach(MIMEText(text))
+
+    # create and attach second required part
+    msg.attach(DeliveryStatusMessage(orig_msg))
+
+    # attach third (optional) part.
+    #
+    # XXX From RFC 6522:
+    #
+    #       When 8-bit or binary data not encoded in a 7-bit form is to be
+    #       returned, and the return path is not guaranteed to be 8-bit or
+    #       binary capable, two options are available. The original message
+    #       MAY be re-encoded into a legal 7-bit MIME message or the
+    #       text/rfc822-headers media type MAY be used to return only the
+    #       original message headers.
+    #
+    #     This is not implemented yet, we should detect if content is 7bit and
+    #     use the class RFC822Headers if it is not.
+#     try:
+#        payload = orig_msg.get_payload()
+#        payload.encode("ascii")
+#    except UnicodeError:
+#        headers = []
+#        for k in orig_msg.keys():
+#            headers.append("%s: %s" % (k, orig_msg[k]))
+#        orig_msg = RFC822Headers("\n".join(headers))
+    msg.attach(orig_msg)
+
+    return msg
+
+
+class BouncerSubprocessProtocol(protocol.ProcessProtocol):
+    """
+    Bouncer subprocess protocol that will feed the msg contents to be
+    bounced through stdin
+    """
+
+    def __init__(self, msg):
+        """
+        Constructor for the BouncerSubprocessProtocol
+
+        :param msg: Message to send to stdin when the process has
+                    launched
+        :type msg: str
+        """
+        self._msg = msg
+        self._outBuffer = ""
+        self._errBuffer = ""
+        self._d = None
+
+    @property
+    def deferred(self):
+        return self._d
+
+    def connectionMade(self):
+        self._d = defer.Deferred()
```

Comment:
I don't know much about twisted protocols, but seeing that the deferred is used just after the initialization of the object `self._d` might not be initializated as `connectionMade` might haven't be called. Could it be better to move this initialization to the `__init__`? Or do I miss something here?